### ISSUE TITLE
it's not possible to run a docker container using a local image

### ISSUE THREAD

**Issue by zvictor** [pulumi/pulumi-docker#148]
### the problem
Pulumi's `docker.Container` requires the images it runs to **always be pulled from a docker registry**, even when the image was built or is already cached in the local machine. 

As a consequence of that, one cannot build-and-run using Pulumi+Docker in development without having to [deploy a Docker Registry server](https://docs.docker.com/registry/deploying/), which is quite overkill when you should be simply running `docker run` instead.

To make things worse, pulumi also requires `username` and `password` to be set when using a registry, thus forcing us to setup authentication in the registry server. That leads us to one more problem. Docker doesn't accept dummy authentication:

> Using self-signed certificates along with basic authentication requires to also trust the certificate into the OS cert store for some versions of docker (https://docs.docker.com/registry/insecure/#use-self-signed-certificates)

Touching the OS cert store kills the whole idea of "write once, run anywhere" provided by Docker, so it's not an option.

### use case
it's a very basic one: I have a nextjs application which I want to run locally and only later on deploy to the cloud. I was hoping that pulumi could be responsible for setting up both environments, as [allured in the blog](https://www.pulumi.com/blog/pulumi-and-docker-development-to-production/).

### code example:

```typescript
const image = new docker.Image(
    `${name}-image`,
    {
      imageName: `${name}-image`,
      build: {
        context: args.context,
      },
      skipPush: true,
    },
  )

const container = new docker.Container(
    `${name}-container`,
    {
        image: image.imageName,
    },
)
```

### logs

```log
I0309 15:40:57.412647    8472 step_executor.go:321] StepExecutor worker(18): applying step create on urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container (preview false)
I0309 15:40:57.412683    8472 registry.go:146] GetProvider(urn:pulumi:dev::yvy::pulumi:providers:docker::default_1_3_0::d7c972e4-dbfe-4d16-89f8-ea4f0b431fd2)
I0309 15:40:57.412718    8472 provider_plugin.go:641] Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container) executing (#props=13)
I0309 15:40:57.412752    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: __defaults={[{attach} {logDriver} {logs} {mustRun} {name} {readOnly} {rm} {start}]}
I0309 15:40:57.412792    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: attach={false}
I0309 15:40:57.412819    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: envs={[]}
I0309 15:40:57.412846    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: image={cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e}
I0309 15:40:57.412870    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: logDriver={json-file}
I0309 15:40:57.412889    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: logs={false}
I0309 15:40:57.412913    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: mustRun={true}
I0309 15:40:57.412938    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: name={cadiz-container-8da3043}
I0309 15:40:57.412956    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: ports={[{map[__defaults:{[{ip} {protocol}]} external:{7000} internal:{5000} ip:{0.0.0.0} protocol:{tcp}]}]}
I0309 15:40:57.413057    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: __defaults={[{ip} {protocol}]}
I0309 15:40:57.413089    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: external={7000}
I0309 15:40:57.413129    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: internal={5000}
I0309 15:40:57.413148    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: ip={0.0.0.0}
I0309 15:40:57.413167    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: protocol={tcp}
I0309 15:40:57.413189    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: readOnly={false}
I0309 15:40:57.413216    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: restart={always}
I0309 15:40:57.413235    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: rm={false}
I0309 15:40:57.413253    8472 rpc.go:70] Marshaling property for RPC[Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container).inputs]: start={true}
I0309 15:40:57.415449    8472 eventsink.go:60] suppress diff ports: old and new don't have the same length
I0309 15:40:57.415515    8472 eventsink.go:63] eventSink::Debug(<{%reset%}>suppress diff ports: old and new don't have the same length<{%reset%}>)
I0309 15:40:57.415936    8472 eventsink.go:60] suppress diff ports: old and new don't have the same length
I0309 15:40:57.415994    8472 eventsink.go:63] eventSink::Debug(<{%reset%}>suppress diff ports: old and new don't have the same length<{%reset%}>)
I0309 15:40:57.416429    8472 eventsink.go:60] suppress diff ports: old and new don't have the same length
I0309 15:40:57.416464    8472 eventsink.go:63] eventSink::Debug(<{%reset%}>suppress diff ports: old and new don't have the same length<{%reset%}>)
I0309 15:40:57.416775    8472 eventsink.go:60] suppress diff ports: old and new don't have the same length
I0309 15:40:57.416801    8472 eventsink.go:63] eventSink::Debug(<{%reset%}>suppress diff ports: old and new don't have the same length<{%reset%}>)
I0309 15:40:57.417098    8472 eventsink.go:60] suppress diff ports: old and new don't have the same length
I0309 15:40:57.417138    8472 eventsink.go:63] eventSink::Debug(<{%reset%}>suppress diff ports: old and new don't have the same length<{%reset%}>)
I0309 15:40:59.162007    8472 provider_plugin.go:1171] provider received rpc error `Unknown`: `Unable to create container with image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: Unable to pull image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: error pulling image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: Error response from daemon: pull access denied for cadiz-image, repository does not exist or may require 'docker login': denied: requested access to the resource is denied`
I0309 15:40:59.162038    8472 provider_plugin.go:1174] rpc error kind `Unknown` may not be recoverable
I0309 15:40:59.162057    8472 provider_plugin.go:671] Provider[docker, 0xc000c0cfc0].Create(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container) failed: Unable to create container with image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: Unable to pull image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: error pulling image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: Error response from daemon: pull access denied for cadiz-image, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
I0309 15:40:59.162090    8472 eventsink.go:87] eventSink::Error(<{%reset%}>Unable to create container with image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: Unable to pull image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: error pulling image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: Error response from daemon: pull access denied for cadiz-image, repository does not exist or may require 'docker login': denied: requested access to the resource is denied<{%reset%}>)
I0309 15:40:59.162183    8472 snapshot.go:300] SnapshotManager: createSnapshotMutation.End(..., false)
I0309 15:40:59.162203    8472 snapshot.go:520] SnapshotManager.markOperationComplete(urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container)
I0309 15:40:59.166291    8472 state.go:191] Saved stack dev checkpoint to: .pulumi/stacks/dev.json (backup=.pulumi/stacks/dev.json.bak)
I0309 15:40:59.166416    8472 step_executor.go:321] StepExecutor worker(18): step create on urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container failed with an error: Unable to create container with image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: Unable to pull image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: error pulling image cadiz-image:255e8439ea169b97a74dc7b410dd175e90073a6047697617cfb63e1de93f143e: Error response from daemon: pull access denied for cadiz-image, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
I0309 15:40:59.166436    8472 step_executor.go:321] StepExecutor worker(18): step create on urn:pulumi:dev::yvy::yvy:docker:Nextjs$docker:index/container:Container::cadiz-container failed, signalling cancellation
I0309 15:40:59.166467    8472 plan_executor.go:244] planExecutor.Execute(...): context finished: context canceled
I0309 15:40:59.166478    8472 step_executor.go:321] StepExecutor worker(-2): worker exiting due to cancellation
```

**Comment by zvictor** [comment]
`@pulumi/docker` is just a wrapper around terraform's docker provider, correct? The [docs for containers in the Docker Terraform Provider](https://www.terraform.io/docs/providers/docker/r/container.html) contain no option to skip the pull of images, even though [in Ansible it's possible](https://docs.ansible.com/ansible/latest/modules/docker_container_module.html#parameter-pull). 
* Does it mean Pulumi can't support that it until Terraform does?
* Is there a way to extend functionalities that don't exist in Terraform?

**Comment by zvictor** [comment]
In Slack @geekflyer has suggested me to:

> Maybe look into using a [Dynamic provider](https://www.pulumi.com/docs/intro/concepts/programming-model/#dynamicproviders) and simply invoke docker commands in that provider (without @pulumi/docker) 

Dynamic providers seem to be a bit too complicated for me as I'm new to Pulumi still, and I'm also afraid it would put me too much "off track".

Still, these are some concerns I have in case I go that direction:
* Has anybody created custom docker commands with Dynamic providers? Can I somehow integrate them with pulumi-docker or would it put me in a bifurcation path?
* Is there any docker + dynamic providers example out there? I would love to have some code example to base my work on instead of starting from scratch.

**Comment by joeduffy** [comment]
> To make things worse, pulumi also requires username and password to be set when using a registry, thus forcing us to setup authentication in the registry server.

We previously had https://github.com/pulumi/pulumi-docker/issues/16, which fixed aspects of this, but I ran into some of this same issuse just last week. For the most part, you can just leave out the `registry` settings altogether; because the image will have the server-name in the tag itself, the `docker push` will succeed without needing to login. I ran into some rough edges here, though, since I was previously logged into another server and had to manually logout first. This sort of workflow with private registries definitely needs some more deep thinking as it's pretty rough today.

**Comment by lukehoban** [comment]
I took a look at this scenario - and I am able to build locally and run containers locally with Pulumi.  Here's what I used:

```ts
import * as docker from "@pulumi/docker";

const image = new docker.Image("mynginx", {
    imageName: "mynginx",
    build: "./app",
    skipPush: true
});

const mynginx = new docker.Container("mynginx", {
    image: image.baseImageName,
    publishAllPorts: true,
    ports: [{ internal: 80 }],
});

export const url = mynginx.ports.apply(d => d && d[0] && d[0].external && `http://${d[0].ip}:${d[0].external}`);
```

And here's what result I get:

```
$ pulumi up --skip-preview
Updating (dev):
     Type                       Name        Status      
 +   pulumi:pulumi:Stack        docker-dev  created     
 +   ├─ docker:image:Image      mynginx     created     
 +   └─ docker:index:Container  mynginx     created     
 
Outputs:
    url: "http://0.0.0.0:32773"

Resources:
    + 3 created

Duration: 3s

Permalink: https://app.pulumi.com/lukehoban/docker/dev/updates/16
```

And ultimately:

```
$ curl $(pulumi stack output url)
<html>
    <head><meta charset="UTF-8">
    <title>Hello, Pulumi!</title></head>
<body>
    <p>Hello, again, local docker!</p>
    <p>Made with ❤️ with <a href="https://pulumi.com">Pulumi</a></p>
</body></html>
```

The one gotcha was that `image.imageName` didn't work as the input to `new docker.Container`.  It appears the format of that, which pins to the specific SHA of the image, is misinterpreted as not being a local image by the `docker.Container` implementation in the upstream provider.  We should look into.  But I believe the scenario here is possible already today.

**Comment by lukehoban** [comment]
There is another slight issue I noticed, which is that the `image` input to `docker.Container` gets normalized to `sha256:<something>`, which means code like above actually recreates the `Container` on every deploy.  Ideally, we would provide an easy way to provide the docker image name (or "id") in this format directly from the output of `docker.Image`.

I can workaround that with this - but its a shame it's needed:

```ts
const imageSHA = image.imageName.apply(imageName => {
    const parts = imageName.split(":");
    return `sha256:${parts[1]}`;
})

const mynginx = new docker.Container("mynginx", {
    image: imageSHA,
    ports: [{ internal: 80, external: 8080 }]
}, { deleteBeforeReplace: true });
```

Which then gives me no-op updates when the local Docker image doesn't change:

```
$ pulumi up --skip-preview
Updating (dev):
     Type                   Name        Status     
     pulumi:pulumi:Stack    docker-dev             
     └─ docker:image:Image  mynginx                
 
Resources:
    3 unchanged

Duration: 2s
```

**Comment by zvictor** [comment]
thank you so much for the investigation, it's very helpful!
I tried a couple of variations of the code you shared and these are my conclusions:

* **Setting `image: image.baseImageName` works, but only if the image previously exists.**

  * Running the code right after a "Reset disk image" throws "Unable to pull image".
  * Running the code again after that works, but I noticed that the container is created while the image is still being recreated, probably using the image of the previous run.

* **the `sha256:${parts[1]}` trick works fine 🎉**
